ASWF/CONTRIBUTING.md

```diff
@@ -0,0 +1,333 @@
+enumerate the rules here.
+
+For the occasional non-clang-format regions of code, NEVER alter somebody
+else's code to reformat just because you found something that violates the
+rules. Let the group/author/leader know, and resist the temptation to change
+it yourself.
+
+Each line of text in your code, including comments, should be at most 80
+characters long. Exceptions are allowed for those rare cases where letting a
+line be longer (and wrapping on an 80-character window) is actually a better
+and clearer alternative than trying to split it into two lines. Sometimes this
+happens, but it's extremely rare.
+
+We prefer three (3) consecutive empty lines between freestanding functions or
+class definitions, one blank line between method declarations within a class
+definition. Put a single blank line within a function if it helps to visually
+separate different sequential tasks, but don't put multiple blank lines in a
+row within a function, or blank lines right after an opening brace or right
+before a closing brace. The goal is to use just enough white space to help
+developers visually parse the code (for example, spotting at a glance where
+new functions begin), but not so much as to spread it out or be confusing.
+
+
+### Identifiers
+
+In general, classes and templates should start with upper case and
+capitalize new words: `class CustomerList;` In general, local variables
+should start with lower case. Macros should be `ALL_CAPS`, if used at all.
+
+If your class is extremely similar to, or modeled after, something in the
+standard library, Boost, or something else we interoperate with, it's ok to
+use their naming conventions. For example, very general utility classes and
+templates (the kind of thing you would normally find in std or boost) should
+be lower case with underscores separating words, as they would be if they
+were standards.
+
+    template <class T> shared_ptr;
+    class scoped_mutex;
+
+Names of data should generally be nouns. Functions/methods are trickier: a
+the name of a function that returns a value but has no side effects should
+be a noun, but a procedure that performs an action should be a verb.
+
+### Class structure
+
+Try to avoid public data members, although there are some classes that serve
+a role similar to a simple C struct -- a very straightforward collection of
+data members. In these, it's fine to make the data members public and have
+clients set and get them directly.
+
+Private member data should be named m_foo (alternately, it's ok to use the
+common practice of member data foo_, but don't mix the conventions within a
+class).
+
+Private member data that needs public accessors should use the convention:
+
+    void foo (const T& newfoo) { m_foo = newfoo; }
+    const T& foo () const { return m_foo; }
+
+Avoid multiple inheritance.
+
+Namespaces: yes, use them!
+
+### Third-party libraries
+
+Prefer C++11 `std` rather than Boost, where both can do the same task.
```

Comment:
What about C++14, 17, etc.? If it's a newer C++, do we use boost for the older C++ and the `std` version for the newer (with #ifdefs)? Having that split would make it easier to drop the boost version once we drop support for the older C++.

Reply:
Haha, I think you found a paragraph that probably dates from when we first were able to use C++11 as a minimum. It should surely be updated. You are right, my preference is to use the boost features only when needed (including proper `#if` guards to select `std` things on a version-by-version basis -- as we do for std::filesystem). We should also mention that a long-term goal is to replace all the current boost use and rid ourselves of that dependency entirely, though there is no specific deadline for doing so.